tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -212,7 +218,383 @@
+   * @param predictions the predictions whose values are in the range <code>[0, 1]</code>.
+   * @param thresholds thresholds in `the range <code>[0, 1]</code>, or {@link #NEG_INF} (used when
+   *     topK is set)
+   * @param topK Optional, indicates that the positive labels should be limited to the top k
+   *     predictions, may be null.
+   * @param classId Optional, limits the prediction and labels to the specified class
+   * @param sampleWeight Optional <code>Tensor</code> whose rank is either 0, or the same rank as
+   *     <code>labels</code>, and must be broadcast to <code>labels</code> (i.e., all dimensions
+   *     must be either <code>1</code>, or the same as the corresponding <code>labels</code>
+   *     dimension).
+   * @param multiLabel indicates whether multidimensional prediction/labels should be treated as
+   *     multilabel responses, or flattened into a single label. When true, the values of <code>
+   *     variablesToUpdate</code> must have a second dimension equal to the number of labels and
+   *     predictions, and those tensors must not be RaggedTensors.
+   * @param labelWeights tensor of non-negative weights for multilabel data. The weights are applied
+   *     when calculating TRUE_POSITIVES, FALSE_POSITIVES, TRUE_NEGATIVES, and FALSE_NEGATIVES
+   *     without explicit multilabel handling (i.e. when the data is to be flattened). May be null.
+   * @param <T> the data type for the variables
+   * @throws IllegalArgumentException If <code>predictions</code> and <code>labels</code> have
+   *     mismatched shapes, or if <code>sampleWeight</code> is not <code>null</code>and its shape
+   *     doesn't match <code>predictions</code>
+   * @return an op to update the given confusion matrix variables.
+   */
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  public static <T extends TNumber> List<Op> updateConfusionMatrixVariables(
+      Ops tf,
+      Map<ConfusionMatrixEnum, Variable<T>> variablesToUpdate,
+      Map<ConfusionMatrixEnum, Assign<T>> varInitializers,
+      Operand<T> labels,
+      Operand<T> predictions,
+      float[] thresholds,
+      Integer topK,
+      Integer classId,
+      Operand<T> sampleWeight,
+      boolean multiLabel,
+      Operand<T> labelWeights) {
+    if (multiLabel && labelWeights != null)
+      throw new IllegalArgumentException(
+          "labelWeights for multilabel data should be handled outside of updateConfusionMatrixVariables when multiLabel is true.");
+
+    if (variablesToUpdate == null || variablesToUpdate.isEmpty()) {
+      return Collections.EMPTY_LIST;
+    }
+
+    Operand<T> lLabels = labels;
+    Operand<T> lPredictions = predictions;
+    Operand<T> lSampleWeight = sampleWeight;
+
+    Operand<TInt32> numThresholds;
+    Operand<TBool> oneThresh;
+    if (multiLabel) {
+      numThresholds = tf.shape.size(lLabels, tf.constant(0));
+      oneThresh = tf.math.equal(tf.constant(1), tf.constant(thresholds.length));
```

Comment:
This is different from the Python implementation, which checks whether the *rank* of thresholds is 1:
~~~Python
    one_thresh = math_ops.equal(
        math_ops.cast(1, dtype=dtypes.int32),
        array_ops.rank(thresholds),
        name='one_set_of_thresholds_cond')
~~~

Reply:
WRT: "This conditional definition of `numThresholds` is different from the Python implementation, which simply reads",
In TF Python, they convert  thresholds to a tensor and get the first dimension value. `Thresholds` may be passed as a `float`, `float[]` or `Operand`.

```
thresholds = ops.convert_to_tensor_v2_with_dispatch(
      thresholds, dtype=variable_dtype)
  num_thresholds = thresholds.shape[0]
```
In Java, thresholds are passed as a `float[]`, not an `Operand`. Creating an Operand and accessing its shape is unnecessary and is equivalent to the taking the` thresholds.length`. Also, because `thresholds` is not an Operand, the rank will always be 1. 

I have not yet run into a case where the `thresholds` are a defined by a multi-dimesional Operand, but maybe we need to support that.

Nonetheless, the code, for now, should take `thresholds.length` instead of dimension 0 of `lLabels`. I will fix.

Reply:
I have changed the thresholds argument to Operand<TFloat32> so now the code reads as:
```
    Operand<TInt32> numThresholds = tf.shape.size(thresholds, tf.constant(0));
    Operand<TBool> oneThresh;
    if (multiLabel) {
      oneThresh = tf.math.equal(tf.constant(1), tf.rank(thresholds));
    } else {
      // TODO handle Ragged Tensors????
      // [y_pred,
      //    y_true], _ = ragged_assert_compatible_and_get_flat_values([y_pred, y_true],
      //                                                   sampleWeights)
      oneThresh = tf.constant(true);
    }
```

Reply:
I had to reshape `numThresholds` to `Shape.scalar()` so that is compatible with the `tf.stack() ` operations later on.
Otherwise, I would get internal TF errors when the operations where evaluated.

```
Operand<TInt32> numThresholds =
        tf.reshape(tf.shape.size(thresholds, tf.constant(0)), tf.constant(Shape.scalar()));
```

```
    Operand<TInt32> threshTilesShape = tf.stack(threshTiles);
    Operand<TInt32> stackedTiles = tf.stack(dataTiles);
```

Reply:
👍 done